server/models/w-user.js

```diff
@@ -0,0 +1,177 @@
+import { app } from '../server'
+import crypto from 'crypto'
+import * as ERRORS from '../lib/error'
+
+module.exports = WUser => {
+  WUser.disableRemoteMethodByName('find') // Disable GET /users
+  WUser.disableRemoteMethodByName('patchOrCreate') // Disable PATCH /users
+  WUser.disableRemoteMethodByName('replaceOrCreate') // Disable PUT /users
+  WUser.disableRemoteMethodByName('prototype.updateAttributes') // Disable PATCH /users/{id}
+  WUser.disableRemoteMethodByName('exists') // Disable HEAD /users/{id}
+  // Will Disable PUT /users/{id} at same time
+  // WUser.disableRemoteMethodByName('replaceById') // Disable POST /users/{id}/replace
+  WUser.disableRemoteMethodByName('createChangeStream') // Disable GET&POST /users/{id}/change-stream
+  WUser.disableRemoteMethodByName('count') // Disable GET /users/count
+  WUser.disableRemoteMethodByName('findOne') // Disable GET /users/findOne
+  WUser.disableRemoteMethodByName('updateAll') // Disable POST /users/update
+  WUser.disableRemoteMethodByName('upsertWithWhere') // Disable POST /users/upsertWithWhere
+
+  const makeSalt = () => Math.round((new Date().valueOf() * Math.random())) + ''
+  const encryptPassword = (password, salt) =>
+    crypto.createHmac('sha1', salt)
+      .update(password)
+      .digest('hex')
+
+  WUser.prototype.hasPassword = function (plain) {
+    if (this.password && plain) {
+      return encryptPassword(plain, this.salt) === this.password
+    }
+  }
+
+  WUser.prototype.createToken = function (ttl, cb) {
+    if (typeof ttl === 'function') {
+      cb = ttl
+      ttl = undefined
+    }
+    const { WToken } = app.models
+
+    WToken.create({
+      ttl: ttl,
+      userId: this.id
+    }, (err, instance) => {
+      cb(err, instance)
+    })
+  }
+
+  WUser.observe('before save', (ctx, next) => {
+    const { WUser } = app.models
+    if (ctx.instance && ctx.instance.user_name && ctx.instance.name) {
+      if (ctx.isNewInstance) {
+        // Create new user
+        WUser.findOne({
+          where: {
+            user_name: ctx.instance.user_name
+          }
+        }, (err, instance) => {
+          if (err) return next(ERRORS.MONGO_ERROR(err))
+
+          if (instance) {
+            // Duplicated user name
+            next(ERRORS.USER_NAME_BEEN_TAKEN)
```

Comment:
yes, this is could be have the TOCTOU problem, to solve the problem, should use the atomic method `findAndModify` of mongodb to make sure the find and create success on same time or fail on same time.